Add VB-LoRA

leo-yangli opened the pull request:
Hi,

I am one of the authors of VB-LoRA: Extreme Parameter Efficient Fine-Tuning with Vector Banks. You can find our paper [here](https://arxiv.org/abs/2405.15179).

This PR integrates our method, VB-LoRA. The implementation and tests follow VeRA and LoRA. I have run pytest locally, and all the tests have passed.

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -1005,7 +1055,7 @@ def test_disable_adapters(self, test_name, model_id, config_cls, config_kwargs):
             rtol, atol = 1e-5, 1e-8
         assert not torch.allclose(outputs_before, outputs_after, rtol=rtol, atol=atol)
-        assert torch.allclose(outputs_before, outputs_disabled)
+        assert torch.allclose(outputs_base, outputs_disabled)
```
If `init_weights` is implemented for VB-LoRA, this can be reverted, right?

leo-yangli replied:
Yes, but we are having difficulty implementing `init_weights` in the same way as LoRA.

leo-yangli replied:
I followed your first suggestion in the comment: 
1. I removed `init_weights`.
2. In tests, I set the vector bank to zero to make VB-LoRA as an identity operation. And initialized the vector bank if training is involved in the test.

I recovered this line back.

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -1072,8 +1124,8 @@ def test_disable_adapters_with_merging(self, test_name, model_id, config_cls, co
-        # check that disabling adapters gives the same results as before training
-        assert torch.allclose(outputs_before, outputs_disabled, atol=atol, rtol=rtol)
+        # check that disabling adapters gives the same results as base model
+        assert torch.allclose(outputs_base, outputs_disabled, atol=atol, rtol=rtol)
```
If `init_weights` is implemented for VB-LoRA, this can be reverted, right?

leo-yangli replied:
Yes, but we are having difficulty implementing `init_weights` in the same way as LoRA.

leo-yangli replied:
Updated.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+        return model
+
+    def test_vblora_parameters(self, mlp):
+        config = VBLoRAConfig(target_modules=["lin0", "lin1", "lin3"], vector_length=2, num_vectors=10)
```
Let's assign `vector_length` and `num_vectors` to variables and use those below in the assertions, so that it's clear where those numbers come from.

leo-yangli replied:
Updated.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+        mlp_vblora = get_peft_model(mlp, config)
+        save_path = tmp_path / "vblora"
+        mlp_vblora.save_pretrained(save_path)
+        print("save_path", save_path)
```
delete

leo-yangli replied:
will do.

leo-yangli replied:
Deleted.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+
+        assert os.path.exists(save_path / "adapter_config.json")
+
+        mlp_vblora_loaded = PeftModel.from_pretrained(mlp, save_path)
```
This is not quite clean: the call to `get_peft_model` above will mutate `mlp`, so this `mlp` here is not really the base model. Let's create a fresh instance of `mlp`.

As you cannot reuse the fixture, you could change the `mlp` fixture to a function `get_mlp` and call this. Or even just initialize `MLP()` directly, as the `mlp` fixture does only that.

leo-yangli replied:
Updated.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+        print("save_path", save_path)
+        assert os.path.exists(save_path / "adapter_config.json")
+
+        mlp_vblora_loaded = PeftModel.from_pretrained(mlp, save_path)
```
Same comment as above regarding re-using `mlp`.

leo-yangli replied:
Updated.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+        input = torch.randn(5, 10)
+        output = mlp_vblora(input)
+        output_loaded = mlp_vblora_loaded(input)
+        assert torch.allclose(output, output_loaded, atol=1e-3, rtol=1e-3)
```
These tolerances are quite generous, can we use smaller ones?

leo-yangli replied:
Agree. will use smaller values.

leo-yangli replied:
Updated.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+        input = torch.randn(5, 10)
+        output = mlp_vblora(input)
+        output_loaded = mlp_vblora_loaded(input)
+        assert torch.allclose(output, output_loaded, atol=1e-3, rtol=1e-3)
```
These tolerances are quite generous, can we use smaller ones?

leo-yangli replied:
Agree. will use smaller values.

leo-yangli replied:
Updated.

BenjaminBossan commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,263 @@
+        if topk <= 0:
+            raise ValueError(f"`topk` {topk} should be a positive integer value")
+
+        if self.in_features % vector_length != 0:
+            raise ValueError(f"`in_features` {self.in_features} must be divisible by `vector_length` {vector_length}")
+        if self.out_features % vector_length != 0:
+            raise ValueError(
+                f"`out_features` {self.out_features} must be divisible by `vector_length` {vector_length}"
+            )
```
It would be great to have tests for this. They can go to `test_initialization.py`.

leo-yangli replied:
Agree. will add tests for this.

leo-yangli replied:
Added a new test.

BenjaminBossan commented on tests/test_vblora.py:
```diff
@@ -0,0 +1,123 @@
+        return X
+
+
+class TestVBLoRA:
```
So the vector bank is shared among all VB-LoRA layers, right? Could we add a test here that checks that this is happening correctly?

leo-yangli replied:
Right! This feature is already tested in line 71. Do you think I need to add a separate test for it?

BenjaminBossan replied:
I see, thanks.

BenjaminBossan commented on src/peft/tuners/vblora/model.py:
```diff
@@ -0,0 +1,403 @@
+                f"Target module {target} is not supported. Currently, only the following modules are supported: "
+                "`torch.nn.Linear`, `transformers.pytorch_utils.Conv1D`."
+            )
+        r = kwargs.pop("r")
```
Why is this necessary?

leo-yangli replied:
I will delete this line.

leo-yangli replied:
Updated.

BenjaminBossan commented on src/peft/utils/save_and_load.py:
```diff
@@ -184,6 +184,27 @@ def renamed_dora_weights(k):
         to_return = {k: state_dict[k] for k in state_dict if "hra_" in k}
+    elif config.peft_type == PeftType.VBLORA:
+        to_return = {}
+        if config.num_vectors < 2**8:
+            indices_dtype = torch.uint8
+        elif config.num_vectors < 2**15:
+            indices_dtype = torch.int16
+        elif config.num_vectors < 2**31:
+            indices_dtype = torch.int32
+        else:
+            indices_dtype = torch.int64
```
How much of a difference does this really make? Could this be omitted to make the code simpler?

leo-yangli replied:
This would be useful when aiming for "extreme" parameter efficiency and aligns with the claims made in the paper.

BenjaminBossan replied:
Okay, is there some measurement how this impacts size? Maybe it's in the paper and I missed it.

leo-yangli replied:
Sure, let's take the Llama2 7B model as an example.

We use a vector bank of 2,048 vectors, each with a length of 256. The rank is set to 4, and K is set to 2, applied to all linear layers. 2,048 vectors can be indexed with an int16.

1. Using int16: The float32-equivalent number of parameters is 680,448, i.e., 2.59 MB.
2. Using int32: The float32-equivalent number of parameters is 758,258, i.e., 2.89 MB. (+11.6% compare to int16)
3. Using int64 (default dtype for index): The float32-equivalent number of parameters is 914,688, i.e., 3.49 MB. (+34.7% compare to int16).

In fact, I believe int16 is sufficient for most cases since it can support up to 32,768 vectors (assuming we don't use negative numbers for indexing). However, I considered all possible cases to ensure robustness. I'd like to know your opinion.

BenjaminBossan commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,263 @@
+        return output_tensor
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
```
I wonder if we can make a sanity check here that users are not trying to train a model that was loaded with `save_topk_weights`. E.g. we could check if `self.training` and if there is -inf in the logits, WDYT? If this check is expensive, it could be done only during the first iteration.

leo-yangli replied:
I added some lines of code to check if both `self.training` and `save_topk_weights` are true.

leo-yangli replied:
I added a sanity check based on your feedback.

BenjaminBossan reviewed (changes requested):
Thanks a lot for this PR that adds VB-LoRA to PEFT. Your implementation already looks quite advanced, is following the existing examples, and includes the required testing, nice :tada: 

In my first review, I focused on the actual VB-LoRA implementation, so I haven't checked the examples and docs yet. As you can see, I added a bunch of comments but those should not be big issues. Please check if they make sense.

Regarding the paper, I also found two typos, in case you want to fix them: "virtrual" and "backpropgatation".

leo-yangli commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,263 @@
+        self.vblora_logits_B[adapter_name] = nn.Parameter(
+            torch.zeros(self.out_features // vector_length, r, num_vectors), requires_grad=True
+        )
+        self.vblora_vector_bank = vblora_vector_bank
```
The `vblora_vector_bank` is initialized and stored in the `VBLoRAModel` class because only one instance of the vector bank is needed. Line 99 gets a reference to this vector bank, so it's important to keep. In the `__init__` function, Line 40 acts like a placeholder and can be removed. WDYT?

BenjaminBossan replied:
Sounds good.

leo-yangli replied:
Updated.

leo-yangli commented on src/peft/utils/save_and_load.py:
```diff
@@ -338,7 +360,30 @@ def set_peft_model_state_dict(
+            PeftType.VBLORA: "vblora_",
         }[config.peft_type]
+        if config.peft_type == PeftType.VBLORA and config.save_topk_weights:
+            b, _ = model.vblora_vector_bank[adapter_name].shape
```
That's right! I'll add a comment.

leo-yangli replied:
Updated.

leo-yangli commented on tests/test_custom_models.py:
```diff
@@ -789,6 +829,8 @@ def test_merge_layers(self, test_name, model_id, config_cls, config_kwargs):
             config_kwargs["init_ia3_weights"] = False
         elif issubclass(config_cls, LNTuningConfig):
             pass
+        elif issubclass(config_cls, VBLoRAConfig):
```
We just want to skip the last else statement `config_kwargs["init_weights"] = False` here but still proceed with the test.

leo-yangli commented on tests/test_custom_models.py:
```diff
@@ -888,7 +933,11 @@ def test_only_params_are_updated(self, test_name, model_id, config_cls, config_k
-        lr = 100.0 if config_kwargs.get("use_dora") and model_id == "EmbConv1D" else 0.5
+        lr = (
+            100.0
+            if config_kwargs.get("use_dora") and model_id == "EmbConv1D" or issubclass(config_cls, VBLoRAConfig)
```
Agree. I will modify the code.

leo-yangli replied:
Updated.

leo-yangli commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,263 @@
+                self.get_base_layer().weight.data -= self.get_delta_weight(active_adapter)
+
+    @staticmethod
+    def _get_low_rank_matrix(logits, vblora_vector_bank, topk):
```
It doesn't have to be a `staticmethod`. I will modify it.

leo-yangli replied:
Updated.

leo-yangli commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,139 @@
+        self.target_modules = (
+            set(self.target_modules) if isinstance(self.target_modules, list) else self.target_modules
+        )
+        if self.save_topk_weights:
```
Agree. I will delete it.

leo-yangli replied:
Deleted.

leo-yangli commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,139 @@
+            )
+        },
+    )
+    init_vector_bank_bound: float = field(
+        default=0.02,
+        metadata={
+            "help": (
+                "The vector bank is initialized with a uniform distribution between -init_vector_bank_bound and"
+                " init_vector_bank_bound."
+            ),
+        },
+    )
+    init_logits_std: float = field(
+        default=0.1,
+        metadata={
+            "help": (
+                "The logits are initialized with a normal distribution with a standard deviation of init_logits_std."
+            ),
+        },
+    )
```
Agree. I will add some comments here.

leo-yangli replied:
Updated.

leo-yangli commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,139 @@
+            )
+        },
+    )
+    save_topk_weights: bool = field(
```
Right. I will modify it.

leo-yangli replied:
Updated.

leo-yangli commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,139 @@
+
+    Paper: https://arxiv.org/abs/.
+
+    Args:
```
Agree. I will add it.

leo-yangli replied:
Updated.

leo-yangli commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,139 @@
+            will not produce the same output as the base model would have without adaptation.
+        modules_to_save (`List[str]`):
+            List of modules apart from Vera layers to be set as trainable and saved in the final checkpoint.
+        init_weights (`bool`):
```
In our approach, we shouldn't initialize the additive adaptor to zero because if the vector bank is initialized to zero, both matrices A and B will be zero, resulting in a zero gradient. As a result, we can't align with other methods regarding `init_weights=True`. Do you have any suggestions in this case?

leo-yangli replied:
I followed your first suggestion in the comment: 
1. I removed `init_weights`.
2. In tests, I set the vector bank to zero to make VB-LoRA as an identity operation. And initialized the vector bank if training is involved in the test.

leo-yangli commented on tests/test_custom_models.py:
```diff
@@ -823,6 +865,9 @@ def test_safe_merge(self, test_name, model_id, config_cls, config_kwargs):
             # LNTuning do not take init_weights
             pass
+        elif issubclass(config_cls, VBLoRAConfig):
+            # VBLoRA do not take init_weights
```
The test proceeds. VB-LoRA doesn't need to set `config_kwargs["init_weights"] = False`.

leo-yangli commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,263 @@
+                if active_adapter not in self.vblora_logits_A.keys():
+                    continue
+
+                vblora_logits_A = self.vblora_logits_A[active_adapter]
```
will do.

leo-yangli replied:
Updated.

BenjaminBossan commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,263 @@
+# Copyright 2023-present the HuggingFace Inc. team.
```
Let's update the dates in all new files to 2024.

BenjaminBossan reviewed (commented):
Thanks for your replies.

Regarding the discussion about `init_weights`, as it came up multiple times, I'll reply here:

From my understanding, the problem is that A and B matrices share the same vector bank. Therefore, if the vector bank is initialized to zeros, both A and B are zeros and therefore there is no gradient. This is unlike LoRA, where only B is zeros.
Also, it's not possible to set the logits for B to zero, as a softmax operation is performed, thus the weights will always sum up to 1.

I have three suggestions:

1. For those cases where we want VB-LoRA to be a no-op, we just manually call `nn.init.ones_(model.base_model.vblora_vector_bank["default"])`. The `init_weights` argument should be removed.
2. We add the option to pass `init_weights=False` (with `True` being default) and in that case initialize the vector bank to zeros. This model is untrainable, so we give a warning that this is option only exists for testing purposes.
3. Add a new parameter that scales the contribution from B, which is a learnable parameter that starts at 0. This way, the vector bank is randomly initialized but VB-LoRA still starts as a no-op. Of course, this would be a deviation from the paper.

I also have a general comment about the docs: Let's make sure to mention that there is an option for VB-LoRA to reduce the size of the file when saving, but that the loaded adapter can only be used for inference after that. Ideally, you could give a numerical example (e.g. for Llama3 8b with parameter ..., the size is reduced from X to Y).

BenjaminBossan commented on src/peft/utils/save_and_load.py:
```diff
@@ -338,7 +360,30 @@ def set_peft_model_state_dict(
+                        topk_weights = torch.cat([topk_weights, 1 - topk_weights.sum(-1, keepdim=True)], dim=-1)
+                        topk_logits = torch.log(topk_weights)
+                    else:
+                        raise ValueError("TopK weights were missing!")
```
Yeah, let's not assume that the users manually edit the save files. If they do, a lot of things can break, we cannot check for all of them.

leo-yangli replied:
I removed the ValueError line.

leo-yangli commented:
Thank you for your valuable feedback! I have updated the code accordingly. 

I also updated the documentation in `docs/source/package_reference/vblora.md`. Please let me know if there is anything else I can improve.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2039). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan commented:
Thanks for the updates @leo-yangli. I haven't reviewed yet but saw that the CI fails because of the following test:

`test_parameters_after_loading_model_110_Vanilla_MLP_4_VBLoRA`

Could you please check?

leo-yangli commented:
Thank you for your feedback. I just fixed that failed test.

leo-yangli commented:
It's ready to be reviewed.

BenjaminBossan commented on examples/sequence_classification/VBLoRA.ipynb:
```diff
@@ -0,0 +1,775 @@
+   "id": "ddfc0610-55f6-4343-a950-125ccf0f45ac",
+   "metadata": {},
+   "source": [
+    "In this example, we fine-tune Roberta on a sequence classification task using VB-LoRA."
```
Let's add a sentence that this notebook is based on `examples/sequence_classficiation/VeRA.ipynb` and users can compare the results.

leo-yangli replied:
Added.

BenjaminBossan commented on examples/sequence_classification/VBLoRA.ipynb:
```diff
@@ -0,0 +1,775 @@
+    "    target_modules=['key', 'value', 'query', 'output.dense', 'intermediate.dense'],\n",
+    "    num_vectors=num_vectors,\n",
+    "    vector_length=vector_length,\n",
+    "    save_only_topk_weights=True,\n",
```
Let's add a comment here that to explain this argument. Using this also means that we cannot resume training from checkpoints if training is interrupted for some reason, right? Let's also mention that.

leo-yangli replied:
Added a comment.

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1170,6 +1171,37 @@ def test_vera_add_second_adapter_with_higher_rank(self):
+        with pytest.raises(ValueError, match=msg):
+            get_peft_model(model, config1)
+
+        config2 = VBLoRAConfig(target_modules=["lin1"], vector_length=vector_length)
```
Let's split this into 2 tests.

leo-yangli replied:
Updated.

BenjaminBossan commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,166 @@
+            The length of the vectors in the vector bank. The length of the vectors should be divisible by the hidden
+            dimension of the model.
+        topk (`int`):
+            K value for topk selection.
```
It would be great if you could explain a bit more: What is the default, when should users decrease or increase this number? We can make the assumption that users are familiar with LoRA and would like to try VB-LoRA, they should not have to read the paper to understand the hyper-parameters.

leo-yangli replied:
Added some explanations.

BenjaminBossan commented on src/peft/tuners/vblora/config.py:
```diff
@@ -0,0 +1,166 @@
+            the target modules manually.
+        save_only_topk_weights (`bool`):
+            Whether to only save the topk weights. Models saved in this mode can be used for merging or inference only,
+            not for resuming training.
```
Let's mention here that the point is to reduce the file size.

leo-yangli replied:
Added.

BenjaminBossan commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,249 @@
+        vblora_logits_B = self.vblora_logits_B[adapter]
+
+        # Check for infinity values when training. If found, training was likely resumed from a `save_only_topk_weights` model.
+        if self.training and vblora_logits_A[0, 0].isinf().any():
```
Nice idea to check only a single row.

leo-yangli replied:
Thank you!

BenjaminBossan commented on src/peft/tuners/vblora/layer.py:
```diff
@@ -0,0 +1,249 @@
+        # Check for infinity values when training. If found, training was likely resumed from a `save_only_topk_weights` model.
+        if self.training and vblora_logits_A[0, 0].isinf().any():
+            raise RuntimeError(
+                "Found infinity values in logits. Ensure training was not resumed from a `save_only_topk_weights` model."
```
Let's add: "**VB-LoRA** logits", otherwise this could be confusing.

leo-yangli replied:
Added.

BenjaminBossan commented on tests/test_common_gpu.py:
```diff
@@ -1572,6 +1573,28 @@ def test_vera_add_new_adapter_does_not_change_device(self, mlp):
+        config = VBLoRAConfig(target_modules=["lin0"], vector_length=2)
+        model = get_peft_model(mlp, config)
+        model = model.cuda()
+        model.lin0.vblora_logits_A.cpu()
```
Let's also check `vblora_logits_B`.

leo-yangli replied:
Sure. Added some lines to check `vblora_logits_B`.

BenjaminBossan reviewed (changes requested):
Thanks a lot for the updates. This is almost good to be merged, I only saw a few tiny issues left. Please take a look.

BenjaminBossan commented:
Just something that came to my mind out of curiosity: Did you try experiments with multiple VB-LoRA adapters (for different tasks) that share the same vector bank? So e.g. train the logits and vector bank on task 1, then freeze the vector bank. Next, create a new VB-LoRA for task 2 that uses the frozen vector bank from task 1 and only trains logits? Probably this won't work well enough for task 2 but it would be really cool if it did to save even more parameters.

leo-yangli commented:
Hi @BenjaminBossan , 

That's a great question! Training in an MTL setting is currently an area of ongoing research for us, and we have tried the exact same idea. Preliminary results indicate that when the two tasks are similar (for example, RTE and MRPC from the GLUE benchmark), the second task can still achieve decent performance with a frozen vector bank, although there is a noticeable drop in performance (e.g., 2%). This performance gap tends to decrease when we alternate training between tasks. However, a more robust training strategy for an MTL setting still needs to be explored. I hope this answered your question.

I updated the code based on your review. Moreover, I added a new feature (count VB-LoRA savable parameters) with tests to the code, and it is ready for review.

Thanks!

leo-yangli commented:
Thanks for pointing out the mistake. I have fixed it.

I've also added a new test, `test_save_with_topk_weights`, to test the weights saved in `save_only_topk_weights=True` mode are as expected.

leo-yangli commented:
Thank you @BenjaminBossan! The code quality has improved a lot thanks to your insightful review.
